File: go/internal/tools/linode_audit_export.go
package tools

import (
	"context"
	"fmt"
	"os"

	"github.com/mark3labs/mcp-go/mcp"

	"github.com/chadit/LinodeMCP/go/internal/audit"
	"github.com/chadit/LinodeMCP/go/internal/config"
	linodev1 "github.com/chadit/LinodeMCP/go/internal/genpb/linode/mcp/v1"
)

// AuditExportAnswer dumps a filtered window of audit events to a temp
// file in JSON, CSV, or NDJSON and answers with the path.
//
// It reads the SQLite store when that sink is enabled, falling back to the
// JSONL log otherwise, and is bounded by max_records so an unbounded range
// never reaches memory.
func AuditExportAnswer(
	ctx context.Context,
	request *mcp.CallToolRequest,
	cfg *config.Config,
) (*mcp.CallToolResult, error) {
	format := request.GetString("format", "")
	ext, _ := exportFileExtension(format)

	query, err := buildExportQuery(request)
	if err != nil {
		return mcp.NewToolResultError(err.Error()), nil
	}

	events, err := audit.ExportEvents(ctx, resolveAuditSQLitePath(cfg), audit.ResolveDefaultAuditDir(), query)
	if err != nil {
		return mcp.NewToolResultError(fmt.Sprintf("failed to read audit log: %v", err)), nil
	}

	path, err := writeExportFile(events, format, ext)
	if err != nil {
		return mcp.NewToolResultError(fmt.Sprintf("failed to write export file: %v", err)), nil
	}

	return MarshalProtoToolResponse(&linodev1.AuditExportResponse{
		Path:        path,
		Format:      format,
		RecordCount: IDToInt32(len(events)),
	})
}

// buildExportQuery translates request parameters into a RecentQuery
// whose Limit carries the resolved max_records cap. Returns an error
// for a malformed since/until timestamp.
func buildExportQuery(request *mcp.CallToolRequest) (*audit.RecentQuery, error) {
	query := &audit.RecentQuery{
		Limit:       resolveMaxRecords(request.GetInt("max_records", 0)),
		Tool:        request.GetString("tool", ""),
		IncludeMeta: request.GetBool("include_meta", false),
	}

	since, err := parseOptionalTime(request.GetString("since", ""))
	if err != nil {
		return nil, fmt.Errorf("invalid 'since' timestamp: %w", err)
	}

	query.Since = since

	until, err := parseOptionalTime(request.GetString("until", ""))
	if err != nil {
		return nil, fmt.Errorf("invalid 'until' timestamp: %w", err)
	}

	query.Until = until

	return query, nil
}

// resolveMaxRecords applies the default and hard cap to a requested
// max_records value. Zero or negative means "use the default".
func resolveMaxRecords(requested int) int {
	if requested <= 0 {
		return audit.DefaultExportMaxRecords
	}

	if requested > audit.MaxExportRecords {
		return audit.MaxExportRecords
	}

	return requested
}

// exportFileExtension maps a format name to its file extension,
// reporting false for an unknown format.
func exportFileExtension(format string) (string, bool) {
	switch format {
	case audit.ExportFormatJSON:
		return "json", true
	case audit.ExportFormatCSV:
		return "csv", true
	case audit.ExportFormatNDJSON:
		return "ndjson", true
	default:
		return "", false
	}
}

// writeExportFile creates a temp file with the format's extension,
// encodes events into it, and returns the path. The file is left in
// place for the user to read; the OS reclaims the temp directory on
// its own schedule.
func writeExportFile(events []audit.Event, format, ext string) (string, error) {
	file, err := os.CreateTemp("", "linode-audit-export-*."+ext)
	if err != nil {
		return "", fmt.Errorf("create temp file: %w", err)
	}

	if err := audit.EncodeEvents(file, events, format); err != nil {
		_ = file.Close()
		_ = os.Remove(file.Name())

		return "", fmt.Errorf("encode export: %w", err)
	}

	if err := file.Close(); err != nil {
		return "", fmt.Errorf("close export file: %w", err)
	}

	return file.Name(), nil
}

File: python/src/linodemcp/tools/linode_audit_export.py
"""Phase 3f audit export tool.

``linode_audit_export`` dumps a filtered window of audit events to a
temp file in JSON, CSV, or NDJSON and returns the path. CapMeta, so
every profile can read it.

The SQLite path arrives through the shared module bridge in
:mod:`linodemcp.tools.linode_audit_summary` (the same path the summary
and health tools read), falling back to the JSONL log when unset.

Mirrors ``go/internal/tools/linode_audit_export.go``.
"""

from __future__ import annotations

import json
import tempfile
from typing import Any

from mcp.types import TextContent

from linodemcp.audit import (
    DEFAULT_EXPORT_MAX_RECORDS,
    MAX_EXPORT_RECORDS,
    RecentQuery,
    encode_events,
    export_events,
    resolve_default_audit_dir,
)
from linodemcp.genpb.linode.mcp.v1 import audit_pb2
from linodemcp.tools.helpers import error_response, parse_optional_time
from linodemcp.tools.linode_audit_summary import audit_sqlite_path
from linodemcp.tools.proto_response import serialize_api_response

_ARG_FORMAT = "format"
_ARG_SINCE = "since"
_ARG_UNTIL = "until"
_ARG_TOOL = "tool"
_ARG_MAX_RECORDS = "max_records"
_ARG_INCLUDE_META = "include_meta"


def audit_export_result(arguments: dict[str, Any]) -> list[TextContent]:
    """Export the filtered window to a temp file and return its path.

    The format's legal set is the contract's rule, so by here the name is one
    the encoder knows. A malformed since/until timestamp answers as a
    tool-result error, the way the Go hook's does.
    """
    export_format = arguments.get(_ARG_FORMAT, "")

    try:
        query = _build_export_query(arguments)
    except ValueError as exc:
        return error_response(str(exc))

    events = export_events(audit_sqlite_path(), resolve_default_audit_dir(), query)

    body = encode_events(events, export_format)

    try:
        path = _write_export_file(body, export_format)
    except OSError as exc:
        return error_response(f"failed to write export file: {exc}")

    payload = {
        "path": path,
        "format": export_format,
        "record_count": len(events),
    }
    result = serialize_api_response(payload, audit_pb2.AuditExportResponse())
    return [TextContent(type="text", text=json.dumps(result, indent=2))]


def _build_export_query(arguments: dict[str, Any]) -> RecentQuery:
    """Translate request arguments into a RecentQuery whose limit carries
    the resolved max_records cap. Raises ValueError on a bad timestamp.
    """
    return RecentQuery(
        limit=_resolve_max_records(arguments.get(_ARG_MAX_RECORDS)),
        since=parse_optional_time(arguments.get(_ARG_SINCE, "")),
        until=parse_optional_time(arguments.get(_ARG_UNTIL, "")),
        tool=arguments.get(_ARG_TOOL, ""),
        include_meta=bool(arguments.get(_ARG_INCLUDE_META, False)),
    )


def _resolve_max_records(requested: object) -> int:
    """Apply the default and hard cap to a requested max_records value.
    A missing, non-integer, or non-positive value uses the default.
    """
    if not isinstance(requested, int) or isinstance(requested, bool) or requested <= 0:
        return DEFAULT_EXPORT_MAX_RECORDS

    return min(requested, MAX_EXPORT_RECORDS)


def _write_export_file(body: str, export_format: str) -> str:
    """Write the encoded body to a temp file named with the format
    extension and return its path. The file is left in place for the
    user to read.
    """
    with tempfile.NamedTemporaryFile(
        mode="w",
        prefix="linode-audit-export-",
        suffix=f".{export_format}",
        delete=False,
        encoding="utf-8",
    ) as handle:
        handle.write(body)
        return handle.name
